### DISCUSSION TITLE
<script setup> should register components by name

### DISCUSSION THREAD

**Discussion by chojnicki** [vuejs/rfcs#273] [RFC Discussions]
Previous comment in RFC pull: https://github.com/vuejs/rfcs/pull/227#issuecomment-788678723

With `<script setup>` it is impossible to use `<component :is="">` with component name as :is.

But it worked fine in v2 and v3 with normal composition api. So I think it should also work with <script setup> syntax and I consider this as a bug. Even if this behavior will be not changed, then it at least should be clarified somewhere in RFC and later docs :)

**Comment by LinusBorg** [comment]
It's explicitly not designed that way: https://github.com/vuejs/rfcs/blob/script-setup-2/active-rfcs/0000-script-setup.md#exposing-components-and-directives

**Comment by fxioi** [comment]
Hi. Did you find a solution for the component name to be used in <script setup>?

**Comment by fxioi** [comment]
a solution:
```typescript
<template>
  <button :class="['text' + size]">size: {{ size }}</button>
</template>

<script lang="ts">
export default { name: "test-btn" };
</script>
<script setup lang="ts">
import { defineProps } from "vue";
import type {PropType} from 'vue'
defineProps({
  size: {
      type: String as PropType<IButtonSize>,
      default: "base"
    }
})
</script>
```

**Reply by chojnicki** [reply]
> a solution:

Unfortunately not for me because I need this for vite-svg, so I do not have control over component file. But other than that, this workaround sounds promising. Maybe I will fork svg plugin to implement this to test that. Will see. Thanks ;)

**Reply by ghost** [reply]
To those reacting with the sad face: this solution is actually an official way of doing what OP needs. It's not some hack https://github.com/vuejs/rfcs/blob/master/active-rfcs/0040-script-setup.md#declaring-additional-options

**Comment by ghost** [comment]
I think this also makes it impossible to get `keep-alive` to work with `script setup` components.

Consider this case from here: https://stackoverflow.com/a/65760822/10706046
```
<router-view v-slot="{ Component }">
  <keep-alive include="MoviesList">
    <component :is="Component" />
  </keep-alive>
</router-view>
```
If you have a MoviesList.vue file the autogenerated name `MoviesList` doesn't work. You have to use https://github.com/vuejs/rfcs/discussions/273?sort=top#discussioncomment-450252

**Reply by ghost** [reply]
Ah, the team foresaw this: https://github.com/vuejs/rfcs/blob/master/active-rfcs/0040-script-setup.md#automatic-name-inference

**Comment by Birdknothing** [comment]
If you’re using Vue Router and have named each route, you can try this:
```javascript
routes.forEach(route => route.component().then(mod=>mod.default.name=route.name));
```
hope this helps 😃

**Comment by gslvly** [comment]
have to use a plugin: 

```js
import { Plugin } from 'vite'
export const setupName = (): Plugin => {
  return {
    name: 'setup-add-name',
    enforce: 'pre',
    transform(code, id) {
      if (!id.endsWith('.vue')) {
        return { code, map: null }
      }
      const match = code.match(/<script\s+.*name=('|")(.+?)('|")/)
      const lang = code.match(/<script\s+.*lang=('|")(.+?)('|")/)
      if (match) {
        return {
          code: code + `<script lang="${lang?.[2]}">export default {name: "${match[2]}"}</script>`,
          map: { mappings: '' },
        }
      }

      return { code, map: null }
    },
  }
}```
